### PR TITLE
Exit when calling `root` command with unrecognized option

### PR DESCRIPTION
**Pull request by vepadulano**
Fixes https://github.com/root-project/root/issues/8850

May still need polishing, will add test if logic is approved.

With this commit, when `root` receives an unrecognized options it prints an error (similar to other linux commands errors) and exits before showing the prompt.

```
$ root --nonexistingoption
Error in <TApplication::GetOptions>: unrecognized option '--nonexistingoption'
Try 'root --help' for more information.
$ cp --nonexistingoption
cp: unrecognized option '--nonexistingoption'
Try 'cp --help' for more information.
```

### CODE REVIEW

**Comment by vepadulano** [comment]
Errors above were due to some options unrecognized by `root` were actually necessary for other callables in the test. Last commit moves the check to `TRint` and also now checks if there are multiple unrecognized options issued by the user:

```
$ root --random -z --nonexistingoption
root: unrecognized option '--random'
root: unrecognized option '-z'
root: unrecognized option '--nonexistingoption'
Try 'root --help' for more information.
```

Also changed to using `std::cerr` for simplicity. Before merging we can discuss the correct output stream

File: core/rint/src/TRint.cxx [hunk 1]
```diff
@@ -146,6 +146,16 @@ TRint::TRint(const char *appClassName, Int_t *argc, char **argv, void *options,
    TApplication(appClassName, argc, argv, options, numOptions),
    fCaughtSignal(-1)
 {
+
+   if (*argc > 1){
+      // Early exit if there are remaining unrecognized options
+      for (auto n = 1; n < *argc; n++){
+         std::cerr << "root: unrecognized option '" << argv[n] << "'\n";
```
Comment (pcanal) [inline core/rint/src/TRint.cxx:153]:
```suggestion
         Error("TRint", "root: unrecognized option '%s'", argv[n]);
```

Comment (vepadulano) [reply core/rint/src/TRint.cxx:153]: For our reference, this change leads to the following output:

```
$: root --random -z --nonexistingoption
Error in <TRint::TRint>: root: unrecognized option '--random'
Error in <TRint::TRint>: root: unrecognized option '-z'
Error in <TRint::TRint>: root: unrecognized option '--nonexistingoption'
Error in <TRint::TRint>: Try 'root --help' for more information.
```

I wonder if this is what we want, given that for some other options to `root` there is already the logic of printing to `stderr`: for example `root --help` is [done here](https://github.com/root-project/root/blob/master/core/base/src/TApplication.cxx#L353-L356)

Comment (pcanal) [reply core/rint/src/TRint.cxx:153]: Fair enough.  so indeed `std::cerr` or `fprintf(stderr,` makes more sense.

**Review by pcanal — APPROVED** [review]

**Comment by vepadulano** [comment]
@eguiraud @Axel-Naumann if you're ok with this PR it is ready to be merged :+1:

File: core/rint/test/TRintTests.cxx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+TEST(TRint, UnrecognizedOptions)
+{
+   // Create array of options
+   int argc{4};
+   char **argv = new char *[argc];
+   argv[0] = const_cast<char *>("-q");
+   argv[1] = const_cast<char *>("-z");
+   argv[2] = const_cast<char *>("--nonexistingoption");
+   argv[3] = const_cast<char *>("-b");
```
Comment (eguiraud) [inline core/rint/test/TRintTests.cxx:16]: might simplify this a lot if it works:

```cpp
std::vector<const char *> v{"-q", "-z", "--nonexistingoption", "-b"};
```

(and the clean-up logic at the end can go away)

File: core/rint/test/TRintTests.cxx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+   // Test the error reported on std::cerr
+   std::cerr << std::flush;
+   // Redirect stderr
+   std::streambuf *old_cerr_streambuf = std::cerr.rdbuf();
+   std::ostringstream stderrstream;
+   std::cerr.rdbuf(stderrstream.rdbuf());
+   // Initializing TRint will delete from argv the recognized options leaving only unrecognized ones.
+   TRint app("App", &argc, argv);
+   // Restore old stderr
+   std::cerr.rdbuf(old_cerr_streambuf);
```
Comment (eguiraud) [inline core/rint/test/TRintTests.cxx:27]: you can use `testing::internal::CaptureStderr` to make this more readable, see e.g. `dataframe_cache.cxx` (we rely on those internal interfaces of googletest in a bunch of places, most importantly in `ROOTUnitTestSupport.h`, so you would not be introducing more coupling than we already have)

**Review by eguiraud — APPROVED** [review]
LGTM although i'm not the code owner :smile:

File: core/rint/test/TRintTests.cxx (hunk 2 above)
Comment (vepadulano) [reply core/rint/test/TRintTests.cxx:16]: The problem is the [TRint constructor](https://github.com/root-project/root/blob/e553d2eba5146f89988b1ab80e97f78a72e2b004/core/rint/src/TRint.cxx#L144) that accepts only a `char**` and the warnings that are generated if one tries to [initialize a `char **`](https://godbolt.org/z/GrT354vjr): ISO C++ forbids converting a string constant to 'char*'

File: core/rint/test/TRintTests.cxx (hunk 3 above)
Comment (vepadulano) [reply core/rint/test/TRintTests.cxx:27]: Ok thanks a lot for the hint! Will rewrite this part of the test

File: core/rint/test/TRintTests.cxx (hunk 2 above)
Comment (eguiraud) [reply core/rint/test/TRintTests.cxx:16]: you can `const_cast` `v.data()`, I think?

Comment (vepadulano) [reply core/rint/test/TRintTests.cxx:16]: Yes that works, I'm unsure as how that will work with the fact that `TRint` will modify the array (eliminating the recognized options and leaving only unrecognized ones). I'll build with the modification and let you know

Comment (eguiraud) [reply core/rint/test/TRintTests.cxx:16]: oh, no that doesn't work at all :sweat_smile: nevermind

**Review by Axel-Naumann — APPROVED** [review]

**Pull request merged**